dali/kernels/common/transpose.h

```diff
@@ -0,0 +1,214 @@
+// Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_KERNELS_COMMON_TRANSPOSE_H_
+#define DALI_KERNELS_COMMON_TRANSPOSE_H_
+
+#include <type_traits>
+#include <utility>
+#include <vector>
+
+#include "dali/core/static_switch.h"
+#include "dali/core/tensor_view.h"
+#include "dali/kernels/common/utils.h"
+
+namespace dali {
+namespace kernels {
+
+namespace detail {
+
+template <typename T, int LevelsLeft, int MaxLevels = -1>
+std::enable_if_t<LevelsLeft == 1> transpose_impl_static(T *dst, const T *src, int max_levels,
+                                                        span<const int64_t> dst_stride,
+                                                        span<const int64_t> src_stride,
+                                                        TensorShape<> size, span<const int> perm) {
+  const auto level = MaxLevels >= 0 ? (MaxLevels - LevelsLeft) : (max_levels - LevelsLeft);
+  auto dst_level_stride = dst_stride[level];
+  auto src_level_stride = src_stride[perm[level]];
+  for (int64_t i = 0; i < size[level]; i++) {
+    *dst = *src;
+    dst += dst_level_stride;
+    src += src_level_stride;
+  }
+}
+
+template <typename T, int LevelsLeft, int MaxLevels = -1>
+std::enable_if_t<(LevelsLeft > 1)> transpose_impl_static(T *dst, const T *src, int max_levels,
+                                                         span<const int64_t> dst_stride,
+                                                         span<const int64_t> src_stride,
+                                                         TensorShape<> size, span<const int> perm) {
+  const auto level = MaxLevels >= 0 ? (MaxLevels - LevelsLeft) : max_levels - LevelsLeft;
+  auto dst_level_stride = dst_stride[level];
+  auto src_level_stride = src_stride[perm[level]];
+  for (int64_t i = 0; i < size[level]; i++) {
```

Comment:
Performance issue: you could use yet another mapping here (I'll call it `dim_order`) which transforms `level`. With that, you could sort the dimensions so that one with the smallest (1) stride is read as the innermost. It would also enable tiling when there are just two levels left.

Reply:
I may try this as a follow-up, but let's get it merged as it is in the simpler form now.